### DISCUSSION TITLE
Rationale for AvoidStringBufferField?

### DISCUSSION THREAD

**Discussion by krallus** [pmd/pmd#3245] [Q&A] [answered]
Hi.  I got the [AvoidStringBufferField](https://pmd.github.io/pmd-6.33.0/pmd_rules_java_bestpractices.html#avoidstringbufferfield) warning when I used StringBuilder in a field, but I am confused by the rationale: "_can grow considerably, and so may become a source of memory leaks if held within objects with long lifetimes_".  Could one not say the same thing about any collection that can grow?  Ultimately, StringBuilder is just a character array.   How is this any different than storing an array or a List in a field?  What are the concerns specific to StringBuilder and similar?

**Comment by adangel** [comment, marked as answer]
>  Could one not say the same thing about any collection that can grow?

That's true.

> How is this any different than storing an array or a List in a field?

There is IMHO no difference.

> What are the concerns specific to StringBuilder and similar?

I guess, that the original author experienced a memory leak and found out that the culprit was a StringBuffer/StringBuilder that was never garbage collected. So he created a specific rule for that case.

But the memory leak/problem could happen with anything - except you use a local variable. Then the lifetime is limited (as long as you don't leak a reference....).

I think, the use case for StringBuilder is actually, that you only create a temporary instance to build up your string and then use "toString()" and work with that. In that case, it doesn't make sense to keep the original StringBuilder around - it only consumes memory (this actually depends on the implementation... e.g. the produced String could share the same char array - until the produced String is garbage collected...). In that sense, storing a StringBuilder in a field is most likely not needed and could point out a implementation problem.

For other classes like Lists, these could lead to similar memory problems. But having a List as a field is probably more common, so it doesn't make sense to flag all fields of type "List". That's just not workable, the rule would be useless then.

Btw.: Using a StringBuilder as field as opposed to using a StringBuffer could potentially raise another error: A field could potentially be accessed by multiple threads - if that is the case for your program, you either need to make sure by other means to synchronize access or use "StringBuffer" - which is thread-safe.

**Reply by oowekyala** [reply]
Maybe, this is meant for eg simple logging frameworks which store everything in memory. Recently we have the problem over in [icyphy/lingua-franca](https://github.com/icyphy/lingua-franca), that a test was failing with OutOfMemoryError because we collect test logs into a StringBuilder.

To me this however sounds like a rule which is too simplistic to be useful. StringBuilder fields are in general perfectly fine, like List fields. It's just in the very specific case that the object has a long lifetime, or other cases pointed out by @adangel, it *could* be a problem. But the rule cannot check that one of these specific cases apply at all. It sounds just like we were prohibiting String fields "because the char array may be big"... Should we retire the rule?

**Reply by krallus** [reply]
I won't get into the gory details, but my use case involves text processing.  I have a class of text processors that are used in much the same way one would use a StringBuilder, but they each have slightly different behaviour and ultimately wrap their own StringBuilder, updating it when they are updated.  I thought nothing of having StringBuilder as a field in this case, but this warning gave me pause.  The most obvious way to avoid the warning is to build a List of strings instead and then use StringBuilder to build the final output when it is requested, but this struck me as unnecessary overhead.  I thought it worthwhile to see if there's something about StringBuilder/StringBuffer that I'm not considering and that would make it different from other collections.  Since that's not the case, I'm just going to disregard the warning.  Thank you both for your replies.

Also, thanks for bringing up the thread safety of StringBuffer vs StringBuilder.  I'll keep that in mind.
